diff --git a/src/resources/gtk/action-bar-end.ui b/src/resources/gtk/action-bar-end.ui
index b5d601fc2..7ed0cca4c 100644
--- a/src/resources/gtk/action-bar-end.ui
+++ b/src/resources/gtk/action-bar-end.ui
@@ -7,6 +7,7 @@
         <property name="transition-type">crossfade</property>
         <child>
           <object class="GtkMenuButton" id="downloads_button">
+            <property name="has_frame">False</property>
             <!-- Translators: tooltip for the downloads button -->
             <property name="tooltip_text" translatable="yes">Downloads</property>
             <property name="receives_default">True</property>
