diff --git a/src/resources/gtk/action-bar-end.blp b/src/resources/gtk/action-bar-end.blp
index 05567f805..9e461d9e0 100644
--- a/src/resources/gtk/action-bar-end.blp
+++ b/src/resources/gtk/action-bar-end.blp
@@ -7,6 +7,7 @@ template $EphyActionBarEnd: Box {
     transition-type: crossfade;
 
     MenuButton downloads_button {
+      has-frame: false;
       tooltip-text: _("Downloads");
       receives-default: true;
 
